File: agentic-dwh/orchestrator.py
"""Claude Pipeline Orchestration Layer (#2 in the diagram).

Top-level controller. Decomposes a user goal into a sequence of agent
invocations and dispatches them. Uses Claude as the planner; each delegate_*
local tool routes to a specialist agent.

Run:
    python orchestrator.py "Build silver+gold for orders, then run DQ checks"
"""
from __future__ import annotations
import asyncio
import sys

from anthropic import Anthropic

from config import load_config
from mcp_bridge import DbtMcpBridge
from agents.base import BaseAgent
from agents import (
    TransformationAgent,
    DataQualityAgent,
    SchemaMigrationAgent,
    IngestionAgent,
    BiInsightAgent,
)


ORCHESTRATOR_SYSTEM = """You are the Claude Pipeline Orchestration Layer for an
agentic data warehouse on Snowflake.

You do NOT execute dbt or SQL directly. You decompose the user's goal and
delegate to specialist agents via the `delegate_*` tools below. Typical flow:

  schema_migration → ingestion → transformation → data_quality → bi_insight

Rules:
  - One delegation at a time. Read each agent's report before deciding the
    next step.
  - If data_quality_agent reports FAIL, route the recommended fix to the
    correct agent (usually transformation) and re-run DQ.
  - Stop when the user's goal is met. Produce a final summary with: what ran,
    what passed/failed, what's deployed in Snowflake, recommended next steps.
  - Never invent results. Quote agent outputs verbatim where it matters.
"""


def build_orchestrator(client: Anthropic, cfg, dbt: DbtMcpBridge) -> BaseAgent:
    # Instantiate specialists once; reuse across delegations.
    specialists = {
        "transformation": TransformationAgent(client, cfg.model, dbt=dbt),
        "data_quality":   DataQualityAgent(client, cfg.model, dbt=dbt),
        "schema_migration": SchemaMigrationAgent(client, cfg.model),
        "ingestion":      IngestionAgent(client, cfg.model),
        "bi_insight":     BiInsightAgent(client, cfg.model, dbt=dbt),
    }

    def make_delegate(key: str):
        async def _delegate(args: dict) -> str:
            task = args.get("task", "")
            if not task:
                return "[error] missing 'task'"
            agent = specialists[key]
            print(f"\n=== Delegating to {agent.name} ===\n{task}\n")
            result = await agent.run(task)
            print(f"\n=== {agent.name} returned ===\n{result}\n")
            return f"[{agent.name} report]\n{result}"
        return _delegate

    schema = {
        "description": "Delegate a sub-task to a specialist agent.",
        "input_schema": {
            "type": "object",
            "properties": {
                "task": {
                    "type": "string",
                    "description": "Self-contained task description for the specialist.",
                }
            },
            "required": ["task"],
        },
    }

    local_tools = {
        f"delegate_{key}": (schema, make_delegate(key))
        for key in specialists
    }

    orchestrator = BaseAgent(
        client=client,
        model=cfg.model,
        dbt=None,                # orchestrator does not call dbt directly
        local_tools=local_tools,
        max_steps=20,
    )
    orchestrator.name = "orchestrator"
    orchestrator.system = ORCHESTRATOR_SYSTEM
    return orchestrator


async def amain(goal: str) -> int:
    cfg = load_config()
    if not cfg.anthropic_api_key:
        print("ERROR: ANTHROPIC_API_KEY not set", file=sys.stderr)
        return 2

    client = Anthropic(api_key=cfg.anthropic_api_key)

    async with DbtMcpBridge(cfg) as dbt:
        print(f"[bootstrap] dbt-mcp tools available: "
              f"{[t['name'] for t in dbt.anthropic_tools()]}")
        orch = build_orchestrator(client, cfg, dbt)
        final = await orch.run(goal)
        print("\n=========== FINAL ===========")
        print(final)
    return 0


def main():
    if len(sys.argv) < 2:
        print('Usage: python orchestrator.py "<goal>"', file=sys.stderr)
        sys.exit(2)
    goal = " ".join(sys.argv[1:])
    sys.exit(asyncio.run(amain(goal)))


if __name__ == "__main__":
    main()

File: agentic-dwh-app/agents/orchestrator.js
"use strict";
const BaseAgent = require("./base");
const {
  TransformationAgent, DataQualityAgent,
  SchemaMigrationAgent, IngestionAgent, BiInsightAgent,
} = require("./specialists");

const ORCHESTRATOR_SYSTEM = `You are the Claude Pipeline Orchestration Layer (Architecture Box 2) — the
central brain for the GOPhER → Snowflake migration (UHC Pharmacy / PharsOnline).

You decompose the goal, route to specialist agents via delegate_* tools, and you
BROKER HUMAN APPROVALS at governance gates. Pipeline order:

  schema_migration → ingestion → transformation → data_quality → bi_insight

Mandatory governance gates (call request_human_approval and wait for the verdict):
  • GATE 1 — before Gold promotion: after transformation builds Silver, request
    approval to promote to Gold (MARTS). Only proceed if APPROVED.
  • GATE 2 — before publishing BI: after BI insights are generated, request
    approval to publish the report. Only finalize if APPROVED.

Rules:
  - One delegation at a time. Read each agent's report before the next step.
  - If data_quality_agent reports FAIL, the pipeline is PAUSED — route the fix to
    transformation_agent, then re-run DQ once.
  - If bi_insight_agent reports a query failure, route it back to
    transformation_agent for auto-correction (Architecture Box 8 feedback loop).
  - If schema_migration_agent reports schema drift, surface it and request approval.
  - If an approval is REJECTED, stop and summarize what was held and why.

GROUND-TRUTH RULE (critical):
  - Every number/table/status in your final summary MUST come from an agent report.
    Never invent figures or claim deployment to a live cloud warehouse — this is an
    in-memory engine over synthetic GOPhER files; describe it accurately.
  - Do not hand-sum totals; quote each agent's reported totals verbatim.
  - For Data Quality, quote the DQ agent's category rollup and STATUS line exactly;
    never add categories or change counts.
  - Treat data as PHI-class; reference the HIPAA-audited governance layer.`;

function buildOrchestrator({ client, model, specialistModel, emit, control }) {
  // Specialists can run on a faster model (e.g. Haiku) while the orchestrator
  // keeps the stronger model for planning + the final report.
  const agentOpts = { client, model: specialistModel || model, emit, control };
  const specialists = {
    schema_migration: new SchemaMigrationAgent(agentOpts),
    ingestion:        new IngestionAgent(agentOpts),
    transformation:   new TransformationAgent(agentOpts),
    data_quality:     new DataQualityAgent(agentOpts),
    bi_insight:       new BiInsightAgent(agentOpts),
  };

  function makeDelegate(key) {
    return {
      schema: {
        description: `Delegate a sub-task to the ${key} specialist agent.`,
        input_schema: { type: "object", properties: { task: { type: "string", description: "Self-contained task." } }, required: ["task"] },
      },
      async handler({ task } = {}) {
        if (!task) return "[error] missing 'task'";
        const agent = specialists[key];
        emit({ type: "delegate", from: "orchestrator", to: agent.name, task });
        const result = await agent.run(task);
        return `[${agent.name} report]\n${result}`;
      },
    };
  }

  const localTools = {};
  for (const key of Object.keys(specialists)) localTools[`delegate_${key}`] = makeDelegate(key);

  /* Human-in-the-loop approval — brokered by the governance layer (server). */
  localTools.request_human_approval = {
    schema: {
      description: "Request human approval at a governance gate (e.g. Gold promotion, BI publish). Blocks until a human approves or rejects.",
      input_schema: {
        type: "object",
        properties: {
          gate: { type: "string", description: "Short gate name, e.g. 'Promote to Gold' or 'Publish BI report'." },
          summary: { type: "string", description: "What the human is approving and the key facts (row counts, KPIs)." },
        },
        required: ["gate", "summary"],
      },
    },
    async handler({ gate, summary } = {}) {
      if (!gate) return "[error] missing 'gate'";
      if (control && control.requestApproval) {
        const decision = await control.requestApproval(gate, summary || "");
        return `Governance gate "${gate}": ${decision.toUpperCase()}` +
          (decision === "approved" ? " — proceed." : " — do NOT proceed; halt and summarize.");
      }
      return `Governance gate "${gate}": APPROVED (no human reviewer attached) — proceed.`;
    },
  };

  const orch = new BaseAgent({ client, model, localTools, maxSteps: 14, maxTokens: 4096, emit, control });
  orch.name = "orchestrator";
  orch.system = ORCHESTRATOR_SYSTEM;
  return orch;
}

module.exports = { buildOrchestrator };
